### PR TITLE
Added fits interface for trim and crop

### LINKED ISSUE #11: Add interface for FITS files
A useful interface to add is direct interaction with FITS files, either through the `FITSIO` types or through loading the file from a string (filename).

## Example usage
```julia
subtracted = subtract_bias("filename.fits", master_bias)
# or even
subtracted = subtract_bias("filename.fits", "master_bias.fits")
```

### FITS files
By dispatching on `FITSIO.ImageHDU` we can access two important pieces of information, the data and the header (metadata)
```julia
# load HDUS
hdus = FITSIO("filename.fits")
hdu = hdus[1] # hdu is a FITSIO.ImageHDU
data = read(hdu)
hdr = read_header(hdu)
```
by having direct access to the header, we can change our interface to allow specifying an "option" using a header value. Traditionally the interface uses strings, but I think we might want to use symbols because it presents a uniform interface that won't interfere with the FITS indices convention.

An example of how we might write this function is
```julia
# use header key
subtract_overscan!(hdu::ImageHDU, key::Symbol; kwargs...) =
    subtract_ovesrcan!(hdu, read_header(hdu)[string(key)]; kwargs...)
# use FITS-style index or Julian indices
subtract_overscan!(hdu::ImageHDU, idxs; kwargs...) =
    subtract_overscan!(read(hdu), idxs; kwargs...)
```
### Filename strings
This can be easily extended to allow just passing in filenames by adding the following functions. Notice how I've added the keyword argument `hdu`, which specifies which HDU is to be used from the fits file (defaulting to 1). This allows easier interoperability of multi-extension FITS files.

```julia
subtract_overscan!(filename::ImageHDU, idxs; hdu=1, kwargs...) =
    subtract_overscan!(FITSIO(filename)[hdu], idxs; kwargs...)
```
notice how I don't specialize the `idxs`, so it could end up being a string, a symbol, or a tuple, or anything. It will get dispatched appropriately!

### Code Generation
Depending on the complexity of the functions, we can avoid having to write functions if they have similar forms by using code generation.
```julia
for func in (:subtract_bias!, :subtract_overscan!, :trimview, :cropview] # etc.
    @eval $func(filename::String, args...; hdu=1, kwargs...) =
              $func(FITS(filename)[hdu], args...; kwargs...)
    @eval $func(hdu::ImageHDU, args...; hdu=1, kwargs...) =
              $func(read(hdu), args...; kwargs...)
end
```
(cc @giordano we _have_ to use `eval` here right?)

Unfortunately this code generation doesn't come with free docs generation, so you'd have to add to the docstring of all the functions something like
```
"""
    cropview(frame, idxs)
    cropview(filename, idxs; hdu=1)
"""
```

### Special Cases
1. function takes two frames
    for `subtract_bias` and `correct_flat` you'll have to add some extra definitions like `func(::String, ::String)`, `func(::AbstractArray, ::String)`

2. Getting indices from file
    some fits files contain information about things like the overscan region in the FITS header- here's an example
	```
	DATASEC =    '[1:2048,1:2068]' / Image area of frame
	CCDSEC  =   '[21:2068,1:2068]' / Image area relative to full chip
	BIASSEC = '[2049:2080,1:2068]' / Overscan area of frame
	```
    so I could reasonably want to be able to use `hdr["BIASSEC"]` directly in `subtract_overscan`, for instance. 
	```julia
    subtracted = subtract_overscan("filename.fits", :BIASSEC)
    # equivalent to
    hdus = FITS("filename.fits")
    frame = read(hdus[1])
    hdr = read_header(hdus[1])
    idxs = hdr[string(:BIASSEC)]
    subtracted = subtract_overscan(frame, idxs)
    ```

### PR DESCRIPTION
**Pull request by siddharthlal25**
Partially fixes #11, will switch to codegen when all code written separately works in a correct manner.

### CODE REVIEW

File: src/methods.jl [hunk 1]
```diff
@@ -207,6 +208,16 @@ julia> trim(frame, "[2:5, 1:5]")
...
 
 
+"""
+    trim(::FITSIO.ImageHDU, idxs)
+    trim(filename, idxs; hdu=1)
+
+Load a FITS file or HDU before trimming.
+"""
+trim(frame::ImageHDU, idxs) = trimview(getdata(frame), idxs)
+trim(frame::String, idxs; hdu = 1) = trim(FITS(frame)[hdu], idxs)
```
Comment (mileslucas) [inline src/methods.jl:218]: please add a method (and relevant documentation) for getting the idxs as a header key. This should be exactly the same as `subtract_overscan`.

**Review by mileslucas — CHANGES_REQUESTED** [review]
For this method, since `trimview` doesn't alter the data, you should do the following:

change
```julia
trim(frame::AbstractArray, idxs) = copy(trimview(frame, idxs))
```
to
```julia
trim(frame, idxs; kwargs...) = copy(trimview(frame, idxs; kwargs...))
```

and then add the methods
```julia
trimview(::AbstractArray, idxs)
trimview(::ImageHDU, idxs)
trimview(::ImageHDU, key::Symbol)
trimview(::String, idxs; hdu=1)
```

**Comment by siddharthlal25** [comment]
`trimview` is a view of the array, and in documentation for `trimview` we have mentioned that subsequent changes to `frame` will lead to change in trimmed version. Whereas in this case when we provide `frame` as `ImageHDU`, mutation (subsequent mutation in `ImageHDU` won't alter the trimmed version) won't take place, that's why I think `trim` would be the right place to define the methods. Thoughts?

Edit:
Also, we shouldn't add any methods to `trimview`, the same as we did for `subtract_overscan!`, because we can't mutate the `ImageHDU` or files as of now.
PS: These are my thoughts, I would like to know your viewpoint.

Edit Again:
Suppose we define `trimview`, we will be using `getdata` to get the data (finally as array), and when we read files (using `read`, it allocates memory because it copies (I think this)), so suppose someone calls `trim` with `ImageHDU` so there will be two copies, one because of this:
```
trim(frame, idxs; kwargs...) = copy(trimview(frame, idxs; kwargs...))
```
and second in the `getdata` in next phase.

Edit Again Again :rofl: 
I should have added the method with `key::Symbol`, I will add them after we resolve this!

**Comment by mileslucas** [comment]
> Also, we shouldn't add any methods to trimview, the same as we did for subtract_overscan!, because we can't mutate the ImageHDU or files as of now.

`trimview` is not a mutating function

> Also, we shouldn't add any methods to trimview, the same as we did for subtract_overscan!, because we can't mutate the ImageHDU or files as of now.

That is a fair point. Thinking a little bit harder my original reasoning behind allowing it is not as strong. So yes, only provide trim functions (this should look _extremely_ similar to subtract_overscan)

**Comment by siddharthlal25** [comment]
I would prefer doing `subtract_bias` and `correct_flat` in separate PR's so that things don't mess up. :smiley: 
Edit: 
Or maybe we can do `subtract_bias` and `correct_flat` in one, and `subtract_dark` in another one. Thoughts?

File: src/methods.jl [hunk 2]
```diff
@@ -276,7 +288,17 @@ julia> crop(frame, (4, 3), force_equal = false)
...
 """
-crop(frame::AbstractArray, shape; kwargs...) = copy(cropview(frame, shape; kwargs...))
+crop(frame, shape; kwargs...) = copy(cropview(frame, shape; kwargs...))
+
+
+"""
+    crop(::FITSIO.ImageHDU, shape; force_equal = true)
+    crop(filename, shape; hdu=1, force_equal = true)
+
+Load a FITS file or HDU before trimming.
```
Comment (mileslucas) [inline src/methods.jl]:
```suggestion
Load a FITS file or HDU before cropping.
```

**Review by mileslucas — CHANGES_REQUESTED** [review]

**Review by mileslucas — APPROVED** [review]

**Review by giordano — APPROVED** [review]

**Pull request merged**